consumer.go

```diff
@@ -80,49 +128,80 @@ func (c *Consumer) isConnected() bool {
 	return c.consumerGroup != nil
 }
 
-// StartListening is a blocking call that tries to establish a connection to Kafka and then starts listening.
-func (c *Consumer) StartListening(messageHandler func(message FTMessage)) {
-	if !c.isConnected() {
-		c.connect()
-	}
+func (c *Consumer) setMonitor(monitor *consumerMonitor) {
+	c.monitorLock.Lock()
+	defer c.monitorLock.Unlock()
 
-	handler := newConsumerHandler(c.logger, messageHandler)
+	c.monitor = monitor
+}
+
+// isMonitorConnected returns whether the consumer monitor is set.
+// It is only set if a successful connection is established.
+func (c *Consumer) isMonitorConnected() bool {
+	c.monitorLock.RLock()
+	defer c.monitorLock.RUnlock()
+
+	return c.monitor != nil
+}
+
+func (c *Consumer) consumeMessages(ctx context.Context, topics []string, handler *consumerHandler) {
+	log := c.logger.WithField("process", "Consumer")
 
 	go func() {
 		for err := range c.consumerGroup.Errors() {
-			c.logger.WithError(err).
-				WithField("method", "StartListening").
-				Error("Error processing message")
+			log.WithError(err).Error("Error consuming message")
 		}
 	}()
 
-	ctx, cancel := context.WithCancel(context.Background())
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Terminating consumer...")
+			return
 
-	go func() {
-		for {
-			if err := c.consumerGroup.Consume(ctx, c.config.Topics, handler); err != nil {
-				c.logger.WithError(err).
-					WithField("method", "StartListening").
-					Error("Error starting consumer")
-			}
-			// Check if context was cancelled, signaling that the consumer should stop.
-			if ctx.Err() != nil {
-				return
+		default:
+			if err := c.consumerGroup.Consume(ctx, topics, handler); err != nil {
```

Comment:
Looking at the example for [ConsumerGroup.Consume](https://pkg.go.dev/github.com/Shopify/sarama#example-ConsumerGroup) I see that they don't reuse the same `ConsumerGroup` when `Consumer` returns an error.

**question:** Shouldn't we do the same ? Do you know what will happen when `Consumer` returns an error and after that we try to `Consume` again ?

Reply:
The consumer group is in fact reused.

If you're referring to the recreation of `exampleConsumerGroupHandler` instances, then I believe that's a personal preference. I guess the example is done in such way because Sarama's documentation explicitly points out that `ConsumerGroupHandler` instances can be called from different goroutines concurrently so it's important to manage their state. However we're thread-safe so there aren't any issues with running it like this.

Reply:
I was actually referring to the `ConsumerGroup` object in the infinite for loop. (Looking at my comment again it is bit confusing.
`err := group.Consume(ctx, topics, handler)` when this err is not nil they don't try to call `Consume` on the same object again.

Reply:
The `panic()` we're seeing there is definitely not an example of how the consumption loop should look like. The `Consume()` function can return an error for a lot of different scenarios which could be temporary issues (e.g. refreshing metadata for topics failing). It is correct to log the error and continue the execution.